src/query/block/column.go

```diff
@@ -133,154 +108,162 @@ func (c *colBlockIter) Next() bool {
+func (it *colSeriesIter) Current() UnconsolidatedSeries {
+	var (
+		bounds = it.meta.Bounds
+		steps  = bounds.Steps()
+		start  = bounds.Start
+		step   = bounds.StepSize
+	)
 
-	r := cb.enforcer.Add(xcost.Cost(len(values)))
-	if r.Error != nil {
-		return r.Error
+	if it.dps == nil {
+		it.dps = make(ts.Datapoints, steps)
 	}
 
-	cb.blockDatapoints.Inc(int64(len(values)))
-	columns[idx].Values = append(columns[idx].Values, values...)
-	return nil
-}
+	for i := 0; i < steps; i++ {
+		it.dps[i] = ts.Datapoint{
+			Timestamp: start,
+			Value:     it.columns[i][it.startIdx+it.idx],
+		}
 
-// AddCols adds the given number of columns to the block.
-func (cb ColumnBlockBuilder) AddCols(num int) error {
-	if num < 1 {
-		return fmt.Errorf("must add more than 0 columns, adding: %d", num)
+		start = start.Add(step)
 	}
 
-	newCols := make([]column, num)
-	cb.block.columns = append(cb.block.columns, newCols...)
-	return nil
+	return UnconsolidatedSeries{
+		datapoints: it.dps,
+		Meta:       it.seriesMeta[it.idx],
+	}
 }
 
-// PopulateColumns sets all columns to the given row size.
-func (cb ColumnBlockBuilder) PopulateColumns(size int) {
-	cols := make([]float64, size*len(cb.block.columns))
-	for i := range cb.block.columns {
-		cb.block.columns[i] = column{Values: cols[size*i : size*(i+1)]}
-	}
+// Next moves to the next item in the iterator. It will return false if there
+// are no more items, or if encountering an error.
+func (it *colSeriesIter) Next() bool {
+	it.idx = it.idx + 1
+	return it.startIdx+it.idx < it.endIdx
 }
 
-// SetRow sets a given block row to the given values and metadata.
-func (cb ColumnBlockBuilder) SetRow(
-	idx int,
-	values []float64,
-	meta SeriesMeta,
-) error {
-	cols := cb.block.columns
-	if len(values) == 0 {
-		// Sanity check. Should never happen.
-		return errors.New("cannot insert empty values")
-	}
+// Err returns any error encountered during iteration.
+func (it *colSeriesIter) Err() error { return nil }
 
-	if len(values) != len(cols) {
-		return fmt.Errorf("inserting column size %d does not match column size: %d",
-			len(values), len(cols))
-	}
+// Close frees up resources held by the iterator.
+func (it *colSeriesIter) Close() { /* no-op*/ }
 
-	rows := len(cols[0].Values)
-	if idx < 0 || idx >= rows {
-		return fmt.Errorf("cannot insert into row %d, have %d rows", idx, rows)
+// SeriesIter returns a series-wise block iterator, giving unconsolidated
+// by series.
+func (c *columnBlock) SeriesIter() (SeriesIter, error) {
+	steps := c.meta.Bounds.Steps()
+	if len(c.columns) != steps {
+		return nil, fmt.Errorf("mismatch in series columns and meta bounds, "+
+			"columns: %d steps: %d bounds: %v", len(c.columns), steps, c.meta.Bounds)
 	}
 
-	for i, v := range values {
-		cb.block.columns[i].Values[idx] = v
-	}
+	return &colSeriesIter{
+		columns:    c.columns,
+		meta:       c.meta,
+		seriesMeta: c.seriesMeta,
+		startIdx:   0,
+		endIdx:     len(c.seriesMeta),
+		idx:        -1,
+	}, nil
```

Comment:
nit: Is it worth extracting this into a `newColSeriesIter(...)` method?  Just so when reading the code it's clear where it's constructed?  I was looking for if `idx` was correctly set to `-1` and just found it hard to find, usually I look for a `newFoo(...)` function vs searching for `&colSeriesIter{...}`.

Reply:
Sounds good